renormalizer/mps/mps.py

```diff
@@ -1240,6 +1240,156 @@ def full_wfn(self) -> np.array:
+            tensor = tensordot(tensor, rtensor, ([-1],[0]))
+            if ms.ndim == 3:
+                tensor = tensordot(tensor, ms, ([0,-1],[0,-1]))
+            else:
+                tensor = tensordot(tensor, ms, ([0,-1,-2],[0,-1,-2]))
+            assert xp.allclose(tensor, tensor.T.conj())
+            rdm[ims] = asnumpy(tensor)
+
+        return rdm
+    
+    def calc_2ordm(self):
+        r""" Calcuate 2-orbital reduced density matrix
+        :math: `\rho_{ij} = \textrm{Tr}_{k \neq i, k \neq j} | \Psi \rangle \langle \Psi |`
+        
+        Returns
+        -------
+        rdm: Dict
+            {(0,1):\rho_{01}, (0,2):\rho_{02}, ...}. The key is a tuple of index of the site.
+        """
+        
+        identity = Mpo.identity(self.mol_list)
+        environ_R = Environ(self, identity, "R")
+        environ_L = Environ(self, identity, "L")
+        L_component = []
+        R_component = []
+        rdm = {}
+        # first construct 1-orbital environment
+        for ims, ms in enumerate(self):
+            ltensor = environ_L.GetLR("L", ims-1, self, identity,
+                    itensor=None, method="Enviro")
+            ltensor = ltensor.reshape(ltensor.shape[0], ltensor.shape[-1])
+            tensor = tensordot(ltensor, ms.conj(), ([0],[0]))
+            if ms.ndim == 3:
+                tensor = tensordot(tensor, ms, ([0],[0]))
+            elif ms.ndim == 4:
+                tensor = tensordot(tensor, ms, ([0,2],[0,2]))
+            L_component.append(tensor.transpose((0,2,1,3)))
+            
+            rtensor = environ_R.GetLR("R", ims+1, self, identity,
+                    itensor=None, method="Enviro")
+            rtensor = rtensor.reshape(rtensor.shape[0], rtensor.shape[-1])
+            tensor = tensordot(ms.conj(), rtensor, ([-1],[0]))
+            if ms.ndim == 3:
+                tensor = tensordot(tensor, ms, ([-1],[-1]))
+            elif ms.ndim == 4:
+                tensor = tensordot(tensor, ms, ([2,-1],[2,-1]))
+            R_component.append(tensor.transpose((0,2,1,3)))
+        
+        # merge two 1-orbital environment together
+        for ims in range(self.site_num):
+            tensor = L_component[ims]
+            for jms in range(ims+1, self.site_num):
+                if jms != ims+1:
+                    kms = jms - 1
+                    tensor = tensordot(tensor, self[kms].conj(), ([2],[0]))
+                    if self[kms].ndim == 3:
+                        tensor = tensordot(tensor, self[kms], ([2,3],[0,1]))
+                    elif self[kms].ndim == 4:
+                        tensor = tensordot(tensor, self[kms], ([2,3,4],[0,1,2]))
+                
+                rtensor = R_component[jms]
+                res = tensordot(tensor, rtensor,
+                        ([2,3],[0,1])).transpose(0,2,1,3)
+                rdm[(ims, jms)] = asnumpy(res.reshape(res.shape[0]*res.shape[1],-1))
+        return rdm
+    
+    def calc_orbital_entropy(self, entropy_type):
```

Comment:
The interface and documentation become confusing considering the old `calc_vn_entropy` code.
How about `calc_entropy(self, entropy_type)` where `entropy_type` can be `"1o"`, `"2o"` or `"bond"`?

Reply:
In quantum dynamics, people may prefer ``DoF``. In QC, it is obscure to use ``DoF`` to denote electrons. The most precise definition here is ``site``. Do you think we should replace it with ``site``?

Reply:
I like `"1siste"` and `"2site"`. 1-site entropy and 2-site entropy are pretty self-explanatory and they look DMRGish.

Reply:
Ok.